keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -60,34 +78,103 @@ class ClusterHandler(
       cloudDriverService.getCluster(resource.spec)
     }
 
-  override fun upsert(resource: Resource<Cluster>, diff: DiffNode?) {
-    val taskRef = runBlocking {
+  private fun Resource<Cluster>.imageIdFromMetadata(): String? =
+    packageResource()?.let { pkg ->
+      imageIdSuppliers.find { it.supports(pkg) }?.imageIdForCluster(pkg, this.spec)
+    }
+
+  interface ImageIdSupplier {
```

Comment:
this is just inline in the class currently but sort of mocks out how I might see making the imageIdForCluster logic pluggable by the resource type for the image